### PR TITLE
Verify ECDSA ( over secp256k1 curve ) Signature

### PR DESCRIPTION
**Pull request by itzmeanjan**
This PR adds necessary routines for implementing ECDSA signature verification. 

> **Note**
> Hashing of input message is computed outside of VM and it's provided as a secp256k1 scalar field element, represented in Montgomery form.

Routine | VM Cycles
--- | --:
`std::crpyto::dsa::ecdsa::secp256k1` | 11,682,007

---

Expected review order #481 -> #482 -> #487 -> `this`

### CODE REVIEW

File: miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs
Comment (vlopes11) [inline miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs:12]: Can we create a constant for this value and document it? Explaining why we picked this point, and what is expected of it

File: miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs
Comment (vlopes11) [inline miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs:13]: Can we leave a comment on why this test is ignored?

**Review by vlopes11 — COMMENTED** [review]
looks good, two unblocking comments

File: miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs
Comment (itzmeanjan) [reply miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs:12]: It's just a random public key, computed from some random private key, using https://github.com/itzmeanjan/secp256k1/blob/37b339db3e03d24c2977399eb8896ef515ebb09b/ecdsa/keygen.py

File: miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs
Comment (itzmeanjan) [reply miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs:21]: Added in https://github.com/maticnetwork/miden/pull/489/commits/a2c90a895ebc8789ca6d2cd70f95a6ffa9fb7f5a

**Comment by itzmeanjan** [comment]
Resolved merge conflicts in https://github.com/maticnetwork/miden/pull/489/commits/a2c90a895ebc8789ca6d2cd70f95a6ffa9fb7f5a

File: miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs
Comment (grjte) [reply miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs:20]: Agreed that it would be good to document this inline.

File: miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs
Comment (grjte) [inline miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs:65]: nit: I would prefer `assert!(test.execute().is_ok());` instead of `unwrap`. I think it's more explicit about what's being asserted in the test.

**Review by grjte — APPROVED** [review]
Exciting to have this implemented, thanks @itzmeanjan! I left one nit inline, but it's not worth blocking the PR over it, so I'll approve and you can update it if you agree with my point.

**Review by vlopes11 — COMMENTED** [review]
non-blocking question about `mac` implementation

File: miden/tests/integration/stdlib/math/secp256k1_scalar_field.rs
Comment (vlopes11) [reply miden/tests/integration/stdlib/math/secp256k1_scalar_field.rs:24]: I tested and its ok as it is

```rust
pub fn mac(a: u32, b: u32, c: u32, carry: u32) -> (u32, u32) {
    let (a, b, c, carry) = (a as u64, b as u64, c as u64, carry as u64);

    // these will never overflow because they are originally `u32`.
    let bc = b * c;
    let acarry = a + carry;

    let result = bc
        .checked_add(acarry)
        .expect("this was the potential problem");

    let hi = (result >> 32) as u32;
    let lo = result as u32;
    (hi, lo)
}

#[test]
fn mac_overflow() {
    let (hi, lo) = mac(u32::MAX, u32::MAX, u32::MAX, u32::MAX);
    assert_ne!(0, hi);
    assert_ne!(0, lo);
}
```

**Review by vlopes11 — APPROVED** [review]

File: miden/tests/integration/stdlib/math/secp256k1_scalar_field.rs
Comment (itzmeanjan) [reply miden/tests/integration/stdlib/math/secp256k1_scalar_field.rs:24]: Thanks @vlopes11

File: miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs
Comment (itzmeanjan) [reply miden/tests/integration/stdlib/crypto/ecdsa_secp256k1.rs:20]: It's already done in https://github.com/maticnetwork/miden/pull/489/commits/a2c90a895ebc8789ca6d2cd70f95a6ffa9fb7f5a

**Pull request merged**
